File: reverseString/reverseString.py
# Create a function that reverses a string

# Solution 1 - Using slicing
""" def reverseString(string):
    return string[::-1] """

# Solution 2 - Using reversed() and join()
""" def reverseString(string):
    return "".join(reversed(string)) """

# Solution 3 - Using a for loop
""" def reverseString(string):
    reversedString = ""
    for i in range(len(string)):
        reversedString += string[len(string) - 1 - i]
    return reversedString """

# Solution 4 - Using a while loop
""" def reverseString(string):
    reversedString = ""
    i = len(string) - 1
    while i >= 0:
        reversedString += string[i]
        i -= 1
    return reversedString """

# Solution 5 - Using recursion
""" def reverseString(string):
    if len(string) == 0:
        return ""
    else:
        return reverseString(string[1:]) + string[0] """

# Solution 6 - Using a stack
def reverseString(string):
    stack = []
    reversedString = ""
    for i in range(len(string)):
        stack.append(string[i])
    for i in range(len(string)):
        reversedString += stack.pop()
    return reversedString


print(reverseString("Hello World!"))
print(reverseString("12345!"))
print(reverseString("Hello World!12345!"))

File: reverseString/reverseString.rb
# Create a function that reverses a string

# Soultion 1 - Using Ruby's built in reverse method
#def reverseString(str)
#    str.reverse
#    end

# Solution 2 - Using a for loop
#def reverseString(str)
#    reversed = ""
#    for i in 0..str.length-1
#        reversed = str[i] + reversed
#    end
#    return reversed
#end

# Solution 3 - Using a while loop
#def reverseString(str)
#    reversed = ""
#    i = 0
#    while i < str.length
#        reversed = str[i] + reversed
#        i += 1
#    end
#    return reversed
#end

# Solution 4 - Using recursion
#def reverseString(str)
#    if str.length == 0
#        return ""
#    else
#        return reverseString(str[1..-1]) + str[0]
#    end
#end

# Solution 5 - Using map

#def reverseString(str)
#    str.split("").map.with_index { |char, i| str[str.length - i - 1] }.join("")
#end

# Solution 6 - Using inject

#def reverseString(str)
#    str.split("").inject { |reversed, char| char + reversed }
#end

# Solution 7 - Using each_char

#def reverseString(str)
#    reversed = ""
#    str.each_char { |char| reversed = char + reversed }
#    reversed
#end

# Solution 8 - Using reduce

def reverseString(str)
    str.split("").reduce { |reversed, char| char + reversed }
end




puts reverseString("Hello World")
puts reverseString("Hello")
puts reverseString("123456789")